R/type_density.R

```diff
@@ -1,237 +1,132 @@
         }
-
-        split_object = lapply(seq_along(split_object), function(ii) {
-            lst = list(
-                x = split_object[[ii]]$x,
-                y = split_object[[ii]]$y,
-                n = split_object[[ii]]$n
+        
+        datapoints = split(datapoints, list(datapoints$by, datapoints$facet))
+        datapoints = Filter(function(k) nrow(k) > 0, datapoints)
+        
+        datapoints = lapply(datapoints, function(dat) {
+            d = density(dat$x, bw = dens_bw, kernel = kernel, n = n)
+            out = data.frame(
+                by = dat$by[1], # already split
+                facet = dat$facet[1], # already split
+                y = d$y,
+                x = d$x
             )
-            if (!is.null(by)) {
-                lst$by = rep_len(by_names[ii], length.out = length(lst$x))
-            } else {
-                lst$by = NULL
-            }
-            if (!is.null(facet)) {
-                lst$facet = rep_len(facet_names[ii], length.out = length(lst$x))
-            } else {
-                lst$facet = NULL
-            }
-            return(lst)
+            return(out)
         })
-        ## combine element by element
-        res = do.call(Map, c(c, split_object))
-        ## now pull out the individual vectors
-        x = res[["x"]]
-        y = res[["y"]]
-        by = res[["by"]]
-        facet = res[["facet"]]
-        bw = sprintf("%.4g", bw)
-        n = res[["n"]]
-        if (is.null(xlab)) {
-            if (length(by_names) > 3 || length(facet_names) > 3) {
-                n = c(n[1:3], "...")
-            }
-            n = paste0("[", paste(n, collapse = ", "), "]")
-            xlab = paste0("N = ", n, "   Joint Bandwidth = ", bw)
-        }
-    }
-    if (type == "area") {
-        ymin = rep(0, length(y))
-        ymax = y
-        # set extra legend params to get bordered boxes with fill
-        legend_args[["x.intersp"]] = 1.25
-        legend_args[["lty"]] = 0
-        legend_args[["pt.lwd"]] = 1
-    }
-
-    ## axes range
-    if (is.null(xlim)) xlim = range(x)
-    if (is.null(ylim)) ylim = range(y)
-
-    ## nice labels and titles
-    if (is.null(ylab)) ylab = "Density"
-    if (is.null(xlab)) xlab = paste0("N = ", object$n, "   Bandwidth = ", sprintf("%.4g", object$bw))
-    if (is.null(main)) main = paste0(paste(object$call, collapse = "(x = "), ")")
-
-    ## facet handling
-    if (!is.null(facet)) {
-        facet = factor(facet, levels = facet_attr[["levels"]])
-        attributes(facet) = facet_attr ## TODO: better solution for restoring facet attributes?
+        datapoints = do.call(rbind, datapoints)
+        datapoints$ymax = datapoints$y
+        datapoints$ymin = rep.int(0, nrow(datapoints))
+        
+        # flags for legend and fill
+        dtype = if (!is.null(bg)) "ribbon" else "l"
+        dwas_area_type = !is.null(bg)
+        
+        out = list(
+            ylab = ylab,
+            xlim = range(datapoints$x),
```

Comment:
Do we need to overwrite `xlim`? What if the user wants to show only a portion of the density? 

I understand we nay need to override `ylim` because `density()` outputs on a different scale, but curious about `x`.

Reply:
And what if users explicitly set `xlim`. Will this break things?